text/src/tail.rs

```diff
@@ -0,0 +1,161 @@
+    /// Validates the command-line arguments to ensure they meet the required constraints.
+    ///
+    /// # Returns
+    /// * `Ok(())` if arguments are valid.
+    /// * `Err(String)` if arguments are invalid, with an error message describing the issue.
+    fn validate_args(&mut self) -> Result<(), String> {
+        // Check if conflicting options are used together
+        if self.bytes.is_some() && self.lines.is_some() {
+            return Err("Options '-c' and '-n' cannot be used together".to_string());
+        }
+
+        if self.bytes.is_none() && self.lines.is_none() {
+            self.lines = Some(10);
+        }
+
+        Ok(())
+    }
+}
+
+/// Prints the last `n` lines from the given buffered reader.
+///
+/// # Arguments
+/// * `reader` - A buffered reader to read lines from.
+/// * `n` - The number of lines to print from the end. Negative values indicate counting from the end.
+
+fn print_last_n_lines<R: BufRead>(reader: R, n: isize) -> Result<(), String> {
+    let lines: Vec<_> = reader.lines().map_while(Result::ok).collect();
+
+    let start = if n < 0 {
+        (lines.len() as isize + n).max(0) as usize
+    } else {
+        (n - 1).max(0) as usize
+    };
+
+    if start > lines.len() {
+        print!("");
+        return Ok(());
+    }
+    for line in &lines[start..] {
+        println!("{}", line);
+    }
+
+    Ok(())
+}
+
+/// Prints the last `n` bytes from the given reader.
+///
+/// # Arguments
+/// * `buf_reader` - A mutable reference to a reader to read bytes from.
+/// * `n` - The number of bytes to print from the end. Negative values indicate counting from the end.
+
+fn print_last_n_bytes<R: Read>(buf_reader: &mut R, n: isize) -> Result<(), String> {
+    let mut buffer = Vec::new();
+
+    buf_reader
+        .read_to_end(&mut buffer)
+        .expect("Failed to read file");
+    let start = if n < 0 {
+        (buffer.len() as isize + n).max(0) as usize
+    } else {
+        (n - 1).max(0) as usize
+    };
+    if start > buffer.len() {
+        print!("");
+        return Ok(());
+    }
+    print!("{}", String::from_utf8_lossy(&buffer[start..]));
+
+    Ok(())
+}
+
+/// The main logic for the tail command.
+///
+/// # Arguments
+/// * `args` - The command-line arguments parsed into an `Args` struct.
+///
+/// # Returns
+/// * `Ok(())` if the operation completes successfully.
+/// * `Err(Box<dyn std::error::Error>)` if an error occurs.
+fn tail(args: &Args) -> Result<(), Box<dyn std::error::Error>> {
+    // open file, or stdin
+    let file: Box<dyn Read> = {
+        if args.file == Some(PathBuf::from("-")) || args.file.is_none() {
+            Box::new(io::stdin().lock())
+        } else {
+            Box::new(fs::File::open(args.file.as_ref().unwrap())?)
+        }
+    };
+    let mut reader = io::BufReader::new(file);
+
+    match args.bytes {
+        Some(bytes) => print_last_n_bytes(&mut reader, bytes)?,
+        None => print_last_n_lines(reader, args.lines.unwrap())?,
+    }
```

Comment:
Weird use of match. `if let ... else if let` could look better here to get rid of the unwrap